### ISSUE TITLE
Central version management

### ISSUE THREAD

**Issue by swharden** [ScottPlot/ScottPlot#1311]
@Kritner from the CodingBlocks slack (https://www.codingblocks.net/slack) has written blog posts ([2021-10-01](https://blog.kritner.com/2021/10/01/dotnet-centralized-package-versioning/), [2018-10-10](https://blog.kritner.com/2018/10/10/using-configuration-files-to-manage-nuget-package-versions-and-other-stuff/)) discussing how to use a single file to define version of multiple projects and their dependencies, making versioning easier to track and update.

## Current System
ScottPlot version and all its dependency package versions are defined in individual csproj files:
* [ScottPlot.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/ScottPlot/ScottPlot.csproj)
* [ScottPlot.WinForms.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.WinForms/ScottPlot.WinForms.NUGET.csproj)
* [ScottPlot.WPF.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.WPF/ScottPlot.WPF.NUGET.csproj)
* [ScottPlot.Avalonia.NUGET.csproj](https://github.com/ScottPlot/ScottPlot/blob/master/src/controls/ScottPlot.Avalonia/ScottPlot.Avalonia.NUGET.csproj)

A versioning-bumping script is manually run to increment ScottPlot version project-wide:
* [Versioner/Program.cs](https://github.com/ScottPlot/ScottPlot/blob/master/dev/tools/Versioner/Program.cs)

## Proposed System
* Use a central file to store ScottPlot version
  * This should be used for `Version`, `AssemblyVersion`, and `FileVersion`
  * Notice that the first is `x.x.x` and the other two are `x.x.x.0` (that fourth number is important)
* Use a central file to store versions of all dependencies _that we want constrained_ (not necessary all packages).
  * ⚠️ Let's be careful which version of `System.Drawing.Common` is used in the core packages. We found that it this package can be _upgraded_ in old solutions, but it breaks them such that they cannot be _downgraded_ again (#1004). For this reason we probably want to keep targeting `4.6.1` for the primary library packages, although tests and demo applications may use newer versions.
* Consider whether to use GitHub Actions to automatically bump the version number before building and publishing NuGet packages in the cloud. See [`.github/workflows/ci.yaml`](https://github.com/ScottPlot/ScottPlot/blob/master/.github/workflows/ci.yaml)

## Potential Drawbacks
* Old versions of Visual Studio may be unable to use the graphical NuGet package manager. I'm not sure about how widely supported central package management is.
  * I think this is okay because it's unlikely typical users working in the source code of this project will need to add NuGet packages.

**Comment by Kritner** [comment]
@swharden should the package version be revved within this PR #1311 to *.25? or is that a part of the build process and/or rollup of several features into a new version?

**Comment by Kritner** [comment]
> Old versions of Visual Studio may be unable to use the graphical NuGet package manager. I'm not sure about how widely supported central package management is.

It is very loosely supported, though honestly I do most of my nuget package management through modifying csproj files directly. I have not looked at the GUI package manager at all using this (experimental) feature that is central package management.

It sure is nice when you have 30+ project files to a repository though 😅

**Comment by swharden** [comment]
I'm re-opening this issue because it's been a few years and central version management support in Visual Studio has probably improved.

I'd like to manage SkiaSharp and Avalonia versions in one place rather than individually editing lots of csproj files, so the strategy here may now be useful to achieve this goal.

**Comment by swharden** [comment]
> It sure is nice when you have 30+ project files to a repository though 😅

When this issue was closed a couple years ago I think the project had 3 project files

This code base now has 39 project files

**Comment by swharden** [comment]
oof, still not quite easy enough to use

Enabling CPM this is an all-or-nothing thing for all projects in the solution

This means if users create new projects (e.g., sandbox apps) they'll get errors which don't clearly point them to the fact that they have to manually edit the `.csproj` file to remove version numbers and add them to the `Directory.Packages.props` file in some other directory.

Also I can't centrally manage just the Avalonia and Skia dependency versions... it looks like just having the `Directory.Packages.props` file requires managing versions of all packages (including things like NUnit dependencies and all that special WinUI stuff), and it gives build errors unless you know how to manually opt-out of central versioning on a package-by-package basis by hand-editing the csproj file

EDIT: in case I'm mistaken, I'll note that I reached this conclusion by adding the central version file and observing 5,000+ build errors with no clear indication of what's wrong -- I find this unacceptably confusing
![image](https://github.com/ScottPlot/ScottPlot/assets/4165489/b0c2bb4c-10d8-4d16-9871-9ec14e95335c)

## Conclusion: I'll pass for now

In conclusion, I still find central version management comes with too much complexity and indirection to make me feel good about using it in an open source project intended to be worked on by developers with a wide variety of experience levels.
